### DISCUSSION TITLE
Beginner Questions

### DISCUSSION THREAD

**Discussion by damccull** [nix-community/kickstart-nix.nvim#30] [Q&A] [answered]
Hello. I have a few beginner questions. If not for having already used the original kickstart.nvim, I would have had a real hard time understanding much about this one with so many extra files. Here's the questions that I think would have helped initially and which I still need to know.

1. How are the files in nvim/plugin actually loaded or included? Is this some default setting somewhere in nvim that I'm missing? I don't see anything in the provided init.lua that explicitly loads them.
2. Without 'lazy', how are these plugins loaded when needed instead of all at startup? Or is it just they all load at startup?
3. Why do plugin configs in nvim/plugin load and error out if the plugin isn't listed in the overlay? Isn't there a way to make those configs be ignored it the plugin itself doesn't exist?

**Comment by mrcjkb** [comment, marked as answer]
Hey :wave: 

> How are the files in nvim/plugin actually loaded or included? Is this some default setting somewhere in nvim that I'm missing? I don't see anything in the provided init.lua that explicitly loads them.

See

- [`:h initialization`](https://neovim.io/doc/user/starting.html#initialization)
- [`:h runtimepath`](https://neovim.io/doc/user/options.html#'runtimepath')

> Or is it just they all load at startup?

With this template, yes. I'm not an advocate for lazy-loading, as [this should be the plugin authors' responsibility, not the users'](https://github.com/nvim-neorocks/nvim-best-practices?tab=readme-ov-file#sleeping_bed-lazy-loading).
If you want a lazy.nvim-like API to implement lazy-loading, which is agnostic of your plugin manager, I have recently released [`lz.n`](https://github.com/nvim-neorocks/lz.n).
But I'm not sure if I want to include it in this template yet.

> Why do plugin configs in nvim/plugin load and error out if the plugin isn't listed in the overlay? Isn't there a way to make those configs be ignored it the plugin itself doesn't exist?

If you want to create a derivation that ignores certain plugin configs, you can use `ignoreConfigRegexes`.

See also:

- [README/Usage](https://github.com/nix-community/kickstart-nix.nvim?tab=readme-ov-file#books-usage)
- [This example](https://github.com/nix-community/kickstart-nix.nvim/blob/52b723c5534bf8227d7123adf6f055ee211c7338/nix/neovim-overlay.nix#L110)

**Comment by damccull** [comment]
Can you explain or point at an example of how to add a plugin (hopefully nix style) that isn't yet in the nixpkgs? I'm interested in using https://github.com/stevearc/conform.nvim but it doesn't seem to have a nixpkg yet. I would want to download and install it as part of the nix process that builds out the nvim derivation. Do I need to make this into a nix module somehow or is there a way to have nvim grab it and install it during the build?

**Reply by mrcjkb** [reply]
[conform.nvim is on nixpkgs](https://search.nixos.org/packages?channel=unstable&from=0&size=50&sort=relevance&type=packages&query=conform-nvim).

You can add the plugins as flake inputs (with `flake = false`) and then use the `buildVimPlugin` function.
[Here's an example](https://github.com/pierrot-lc/nvim-nix).

**Reply by damccull** [reply]
Thanks, this is working great to install the plugins. I have another issue with it though. It seems to be modifying _all_ bash scripts included in the plugin source to set their interpreter to a specific nix store bash path, by hash. This is causing remote-nvim to break on the remote end, as I believe it copies a local bash script to the remote and attempts to execute it.

The remote in my case, being an ubuntu container, does not have a nix store and therefore can't execute its version of bash to run the script. Is this something you can solve, or do you have a link to the repo that maintains the vimUtils? I can't seem to find it. I assume it should instead be looking for 'env bash' for the shebang rather than an explicit path, but either way I would like to find a way to omit specific scripts from mangling if possible.

**Reply by mrcjkb** [reply]
That's just how Nix works. 
If you want to use a derivation on a remote machine, you can use `nix copy`.

https://determinate.systems/posts/moving-stuff-around-with-nix/

**Reply by damccull** [reply]
Well, that's fine and all, but the remote-nvim plugin doesn't do that. Its goal is a fresh nvim install on the remote to act as a headless neovim instance to connect to, I believe.

It tries to install a fresh nvim over on the remote via a script. That script never gets run on my local system, so it doesn't need to be nixish unless, I suppose, someone wanted to run it on a nix remote. Is there no way to override or list specific files to be ignored for changes while building?

**Reply by mrcjkb** [reply]
Oh, it's a plugin you're using?

If you're getting the plugin from nixpkgs, I would suggest opening an issue there.

Otherwise, you can override the plugin derivation (using `overrideAttrs`) or build it manually.

**Reply by damccull** [reply]
Yes, it is a plugin called '[remote-nvim](https://github.com/amitds1997/remote-nvim.nvim)'. It's not in nix-packages so I had to use the vimTools.buildVimPlugin function you recommended to make it work. Where do I learn about the overrides or build process for this plugin? Sorry for my newbiness. I'm happy to do the learning, but Nix documentation on the internet is, in my experience, difficult to find current, relevant information on weird topics like this.

**Reply by mrcjkb** [reply]
Actually, you can just add `dontPatchShebangs = true;` to the attrset you pass to  `buildVimPlugin`.

**Reply by damccull** [reply]
Ah, perfect! Can you tell me where you found that? I couldn't even seem to find a location where buildVimPlugins comes from.

**Reply by mrcjkb** [reply]
`buildVimPlugin` is just a wrapper around `stdenv.mkDerivation`.
 
Here's the relevant documentation: https://nixos.org/manual/nixpkgs/stable/#variables-controlling-the-fixup-phase

**Comment by damccull** [comment]
That's perfect. Thank you.
